xla/service/algebraic_simplifier.cc

```diff
@@ -519,6 +484,53 @@ int64_t GetReduceFlops(const HloInstruction* reduce) {
 
 }  // namespace
 
+bool AlgebraicSimplifierVisitor::IsPositive(
+    const HloInstruction* hlo, const AlgebraicSimplifierOptions& options) {
+  // Utility only handles real types.
+  if (IsAnyOperandComplex(hlo)) {
+    return false;
+  }
+  switch (hlo->opcode()) {
+    case HloOpcode::kGetTupleElement: {
+      const HloInstruction* gte_operand = hlo->operand(0);
+      switch (gte_operand->opcode()) {
+        case HloOpcode::kCustomCall: {
+          const auto& target = gte_operand->custom_call_target();
+          return target ==
+                     options.get_cudnn_batchnorm_forward_training_metadata() &&
+                 hlo->tuple_index() == 2;
+        }
+        default:
+          return false;
+      }
+    }
+    case HloOpcode::kExp:
+      return true;
```

Comment:
Exp can be zero if the argument is negative infinity. Similarly, for the added `kMultiply` case where the LHS and RHS are positive, the result can underflow to zero.

Hmmm, but I see existing cases have similar problems, even without this PR. E.g. kPower can also be zero if the RHS is negative infinity. Perhaps we don't care about cases where underflow happen, or where arguments are infinity? @akuegel, do you have an opinion here?

Reply:
Sorry, I somehow missed this question. Good point, I think we should not return true for kExp. And same for kPower. We can return true for them in the IsNonNegative function.
Also, good catch regarding underflow to zero, I think we should also not return true for kMultiply.